gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/VectoredIOImpl.java

```diff
@@ -0,0 +1,382 @@
+   * @param allocate Function to allocate ByteBuffer for reading.
+   * @param gcsFs GCFS implementation to use while creating channel and reading content for ranges.
+   * @param fileInfo FileInfo of the gcs object agaisnt which range request are fired, this can be
+   *     null for some code path fall back to URI path provided.
+   * @param gcsPath URI of the gcs object for which the range requests are fired.
+   * @throws IOException If invalid range is requested, offset<0.
+   */
+  public void readVectored(
+      List<? extends FileRange> ranges,
+      IntFunction<ByteBuffer> allocate,
+      GoogleCloudStorageFileSystem gcsFs,
+      FileInfo fileInfo,
+      @Nonnull URI gcsPath)
+      throws IOException {
+    ReadChannelProvider channelProvider = new ReadChannelProvider(gcsFs, fileInfo, gcsPath);
+    readVectored(ranges, allocate, channelProvider);
+  }
+
+  private void readVectored(
+      List<? extends FileRange> ranges,
+      IntFunction<ByteBuffer> allocate,
+      ReadChannelProvider channelProvider)
+      throws IOException {
+    List<? extends FileRange> sortedRanges = validateNonOverlappingAndReturnSortedRanges(ranges);
+    for (FileRange range : ranges) {
+      // TODO(user): upgrade to use validateAndSortRanges
+      validateRangeRequest(range);
+      CompletableFuture<ByteBuffer> result = new CompletableFuture<>();
+      range.setData(result);
+    }
+    if (shouldMergeRanges(ranges)) {
+      updateRangeSizeCounters(sortedRanges.size(), sortedRanges.size());
+      // case when ranges are not merged
+      for (FileRange sortedRange : sortedRanges) {
+        long startTimer = System.currentTimeMillis();
+        boundedThreadPool.submit(
+            () -> {
+              logger.atFiner().log("Submitting range %s for execution.", sortedRange);
+              readSingleRange(sortedRange, allocate, channelProvider);
+              long endTimer = System.currentTimeMillis();
+              storageStatistics.updateStats(
+                  GhfsStatistic.STREAM_READ_VECTORED_READ_RANGE_DURATION,
+                  endTimer - startTimer,
+                  channelProvider.gcsPath);
+            });
+      }
+    } else {
+      List<CombinedFileRange> combinedFileRanges = getCombinedFileRange(sortedRanges);
+      updateRangeSizeCounters(sortedRanges.size(), combinedFileRanges.size());
+      // case where ranges can be merged
+      for (CombinedFileRange combinedFileRange : combinedFileRanges) {
+        CompletableFuture<ByteBuffer> result = new CompletableFuture<>();
+        combinedFileRange.setData(result);
+        long startTimer = System.currentTimeMillis();
+        boundedThreadPool.submit(
+            () -> {
+              logger.atFiner().log("Submitting combinedRange %s for execution.", combinedFileRange);
+              readCombinedRange(combinedFileRange, allocate, channelProvider);
+              long endTimer = System.currentTimeMillis();
+              storageStatistics.updateStats(
```

Comment:
Also, we update the stats before the operation so that it will track it even if there is some exception.